fix org.apache.camel.quarkus.component.lra.it.LraTest failed #2523

ffang opened the pull request:
[ ] An issue should be filed for the change unless this is a trivial change (fixing a typo or similar). One issue should ideally be fixed by not more than one commit and the other way round, each commit should fix just one issue, without pulling in other changes.
[ ] Each commit in the pull request should have a meaningful and properly spelled subject line and body. Copying the title of the associated issue is typically enough. Please include the issue number in the commit message prefixed by #.
[ ] The pull request description should explain what the pull request does, how, and why. If the info is available in the associated issue or some other external document, a link is enough.
[ ] Phrases like Fix #<issueNumber> or Fixes #<issueNumber> will auto-close the named issue upon merging the pull request. Using them is typically a good idea.
[ ] Please run mvn process-resources -Pformat (and amend the changes if necessary) before sending the pull request.
[ ] Contributor guide is your good friend: https://camel.apache.org/camel-quarkus/latest/contributor-guide.html

jamesnetherton commented on integration-tests/lra/src/test/java/org/apache/camel/quarkus/component/lra/it/LraTestResource.java:
```diff
@@ -36,8 +42,16 @@
     @Override
     public Map<String, String> start() {
         try {
+            String hostname = InetAddress.getLocalHost().getHostName();
```
Small nit - I don't see `hostname` used for anything so I guess it can be removed?

ffang replied:
Hi @jamesnetherton ,

Thanks for all the feedback!
I sent a new commit which removed unused code you mentioned above.

Freeman

jamesnetherton commented:
Asside from the `LraTestResource` mods, I'm a bit confused by the changes to this test.

The general idea is that `CreditService` has a limit of 100 credits. The test then makes 3 'orders' amounting to 95 credits. It then attempts another order with a total of 20 to trigger the 'Insufficient Credit' scenario. So from what I can see, the assertions are valid and correct (I think I stole this idea from the Camel main project BTW).

jamesnetherton commented:
Another thought is that, without host networking, you may not be able to run this test properly. The LRA container needs the ability to 'call back' to the test application running on the host.

ffang commented:
> Asside from the `LraTestResource` mods, I'm a bit confused by the changes to this test.
> 
> The general idea is that `CreditService` has a limit of 100 credits. The test then makes 3 'orders' amounting to 95 credits. It then attempts another order with a total of 20 to trigger the 'Insufficient Credit' scenario. So from what I can see, the assertions are valid and correct (I think I stole this idea from the Camel main project BTW).

OK, now I understand better about the testcase.

So the "java.lang.IllegalStateException: Insufficient credit" is expected, but this test still has the logic error.
Since it always creates new orders before checking the credit, so when checking valid order, the expected return value is 4 instead of 3. I adjusted the camel route definition(put checking credit before creating new order) to make the test pass.

Actually this problem also exists in the test counterpart(LRACreditIT#testCreditExhausted)  of camel-lra component, so I think we also need a change in camel project.

ffang commented:
> Another thought is that, without host networking, you may not be able to run this test properly. The LRA container needs the ability to 'call back' to the test application running on the host.

You raised a very good question here which pushed me to went through a bunch docker network documents to give the answer.

You must be on the linux machine so that withNetworkMode(host)(equals --network host, which means container shares the same network stack as host). However, this doesn't work on mac. Per the document here
https://docs.docker.com/network/host/
```
The host networking driver only works on Linux hosts, and is not supported on Docker Desktop for Mac, Docker Desktop for Windows, or Docker EE for Windows Server.
```
Moreover, when "--network host" is used, the "port forwarding"(-p container_port:host_port) related configuration will be ignored, which means for developer on mac(also on windows, if there is any), can't access the lra service in the container from the host. Since we can't guarantee all developers to use linux machine, I think we need the test runnable on major OS.
So I think we should use default network mode(bridge)

From this test, I don't see how LRA container needs the ability to 'call back' to the test application running on the host, since it works for me now on my mac, please test it with your machine if you are on linux. If in any case we need call back from container to host, with default network mode, we can also do it(using hostname docker0 for linux and host.docker.internal for mac and windows)

I think my new commit addressed all concerns we have so far, please take another look.

Thanks!
Freeman

jamesnetherton commented:
> the expected return value is 4 instead of 3

I believe it should be 3. We attempt to create 4 orders, but one will end up as cancelled and the LRA will be rolled back. If you're seeing 4 as the result, it means the compensation `direct:cancelOrder` route is not being invoked. That's why the LRA container needs access to the host application, because it triggers the compensation actions.

ffang commented:
> > the expected return value is 4 instead of 3
> 
> I believe it should be 3. We attempt to create 4 orders, but one will end up as cancelled and the LRA will be rolled back. If you're seeing 4 as the result, it means the compensation `direct:cancelOrder` route is not being invoked. That's why the LRA container needs access to the host application, because it triggers the compensation actions.

Thanks @jamesnetherton  for the quick reply!

Yes, I got your point.
Instead of using localhost for the callback host, we need to use host.docker.internal for mac, and docker0 for linux machine
We need change like
```
-                    String.format("http://localhost:%s", System.getProperty("quarkus.http.test-port", "8081")));
+                    String.format("http://%s:%s", "host.docker.internal",
+                            System.getProperty("quarkus.http.test-port", "8081")));
```
I will push another commit.
Freeman
